### PR TITLE
Adding support for overriding the manifest with sessionStorage and/or localStorage

### PR DESCRIPTION
**Pull request by joeldenning**
The reason why this will be helpful is for testing things in an environment where you want to use a manifest by default, but override just specific services with your own url.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,9 +5,9 @@
   "main": "dist/sofe.js",
   "scripts": {
     "test": "karma start",
-    "build": "babel src --out-dir lib && node build.js",
+    "build": "npm run clean && babel src --out-dir lib && node build.js",
     "rollup": "rollup src/index.js -f cjs -o dist/sofe.js",
-    "clean": "rm -r lib",
+    "clean": "rm -rf lib",
```
Comment (joeldenning) [inline package.json]: `f` so that there is no error when the lib directory doesn't exist.

File: package.json [hunk 2]
```diff
@@ -5,9 +5,9 @@
   "main": "dist/sofe.js",
   "scripts": {
     "test": "karma start",
-    "build": "babel src --out-dir lib && node build.js",
+    "build": "npm run clean && babel src --out-dir lib && node build.js",
```
Comment (joeldenning) [inline package.json]: `clean` before `build` because deleted files in the `src` directory can still exist in the `lib` directory unless you clean first

**Comment by nhumrich** [comment]
This is awesome!

**Comment by blittle** [comment]
Does this need to be in the core library? You can easily do this currently:

```javascript
System.config({
      sofe: {
        manifest: JSON.parse(window.localStorage.getItem("sofe:overrides")),
        manifestUrl: root + '/test/manifests/simple.json'
      }
    });
```

**Comment by blittle** [comment]
The locally defined manifest overrides what is defined in the remote manifest.

**Comment by joeldenning** [comment]
The use case is one that Nick and I discussed a few days ago: what if you want to test your local changes to a service right on an integration/staging/production environment? This would allow you to do that very easily.

Non-Canopy use cases are similar: if a service owner knows of a few production apps that are using his/her service, he could go to those production apps and test out his changes before he deploys the service for reals.

I see it as a great way for service owners to QA things before they deploy.

**Comment by blittle** [comment]
I understand the use case. But again, an application using sofe will usually have to have some configuration passed with `System.config`. Rather than having an implied usage of local storage, if the consumer wanted the above use case, they would just do the following for their application:

```javascript
const overrides = window.localStorage.getItem('sofe:overrides');

System.config({
    sofe: {
        manifest: overrides ? JSON.parse(overrides) : {},
        manifestUrl: "some/address.json"
    }
});

System.config({
    sofe: {
        manifest: { ...someOtherManifest, ...overrides }
    }
});
```

I am just unsure we need the added complexity for a use case that can be solved with an extra two lines of code on the outside.

**Comment by joeldenning** [comment]
Are you more concerned about the complexity to sofe's code, or the implied behavior of `sofe:...` keys in local storage?

I think an analogous precedent to doing something like this is with CLIs. For example, the babel cli has a configuration file, but it can be overridden by a command line option. The reason for the override is because it is not easy to modify .babelrc (with bash code) before you call babel. And also that maybe you don't want to change the .babelrc file, since that is version controlled and the change you're making is just a one-off thing that isn't going to get checked in.

Likewise, in a use case where the sofe service owner *does not own the application*, it is not easy for him to change the SystemJS config after the config.js is loaded in the browser, but before the application bootstraps. If the service owner *did own the application*, then I would agree with you that this pr is frivolous. However, since one of the driving ideas behind sofe is "it makes it easier to scale organizations and teams", I think it makes sense to cater to a use case where a service owner does not also own all the apps that use the service.

Can you think of a better way to test a service when you don't have access to the source code of the app that uses it?

**Comment by blittle** [comment]
Thinking about it more, it might be really cool if we had a chrome extension `sofe-inspector`. The extension would tell you what the currently available services are and allow you to override them. Maybe having conditional overrides too, ie. `window.location.hostname === 'app-local.canopy.dev'`

If we did that, we would want control over how things are saved to local storage. But maybe we'd want a different data structure saved. Maybe a json structure saved under the localstorage key `sofe:overrides`:

```json
[
  {
    "conditional": "window.location.hostname === 'app-local.canopy.dev'",
    "sofe": {
      "manifest": { },
      "manifestUrl": ""
    } 
  }
]
```

Then we could easily extend the sofe config properties with the overrides if the conditional evaluates truthy. Making this work well would be dependent on an extension with a nice UI.

**Comment by blittle** [comment]
Didn't see your above response. Good point on the use case of not being in control of the root application source.

**Comment by blittle** [comment]
The extension could also provide warnings like you guy suggested earlier.

**Comment by joeldenning** [comment]
Interesting idea about the conditional loading. I agree that it would require some type of UI to manage it though. Also, I feel like doing complex conditional loading based on local storage may make it easy for the user to the wrong thing -- imo, you shouldn't put code into local storage to be `eval`'ed later on. Especially in a production setting.

In regards to the Chrome extension idea, I like it. Things would definitely be simplified. However, we would need to make sure that it is "easy enough" for non-chrome users to accomplish the same things without too much pain. Which is why I settled on a very basic local storage format, that you can use with or without a chrome extension.

**Comment by joeldenning** [comment]
About the warning you mentioned, what do you think of the following messaging being logged whenever a local storage override is being used:

`Using local storage override to load sofe service ${serviceName}. To remove this override, run the following command: window.localStorage....`

**Comment by blittle** [comment]
The main use case I was thinking about for conditional overrides was to turn things on based upon domains, but that is already taken care of with local storage being scoped to domains. Lets keep the format you already have here. It keeps things simple. I like that warning message. A good default built into sofe itself. Perhaps the extension could give a more visual indicator without the console being loaded.

**Comment by joeldenning** [comment]
Cool, I will add in the logging statements

**Comment by blittle** [comment]
Want to also add something about it to the readme?

**Comment by joeldenning** [comment]
Sure thing

**Comment by blittle** [comment]
I'll wait to merge https://github.com/CanopyTax/sofe/pull/4 until this is merged, likely to be some conflicts.

**Comment by joeldenning** [comment]
I used `console.log` instead of `console.warn` because I feel like more people will read it if it's black text instead of yellow text. Willing to change it though.

**Pull request merged**